Kinsey's Rotation Spring 2025


TF_list:
    - original_data:
        - known_ADs_considering_isoforms_and_canonical.csv: Sanjana's list of ADs 
        - lambert_supp_tables.csv: TFs from Lambert et al.
        - proteinatlas-TFs.csv: TF-annotated proteins from Human Protein Atlas
        - tycko_soto_delrosso_RD_coordinate_data.csv: Lab list of RDs
    - TF_AD_RD.ipynb:
        - Annotating AD/RD/both/none for all TFs
    -   TF-specificity.ipynb:
        - scRNA-seq cell-specificity and co-expression analysis on ADs/RDs
        - Domain/motif/compbias annotations for ADs/RDs
    - tfsmapped-gencodes.csv: Uniprot mapping for TFs from cofactors_humanproteinatlas>TF-AD-RD>tfs_combined.csv
    - lambert_tfs.csv: EnsemblID and gene name for TFs in Lambert et al.
    - expression_data: tissue expression data 
    - maps: Uniprot mapping for ADs, RDs, TFs

cofactor_complexes: 
    - data: manually curated csv files of cofactor subunits and paralogs
    - cofactor_expression: datasets of tissue specifity of cofactor paralogs
    - cofactors_old: analysis with non-normalized scRNA data
    - cofactorcomplexes_mining.ipynb: 
        - API calling for ENSG ID to Gtex gene code
        - API calling for Gtex gene code to tissue expression profile from Gtex portal
    - cofactor_analysis.ipynb: 
        - Clustering heatmaps of gene expression by tissue
        - Expression correlation between own-complex paralogs
        - KMeans and Agglomerative Clustering of subunits and tissue types
        - BAF-specific exploration in brain tissues
    - cofactor_analysis_normalized.ipynb:
        - Same as above, but with normalized gene expression values across gene

cofactors_humanproteinatlas:
    - data:
        - TF-AD-RD: directory containing .csv and .tsv files with ADs, RDs, and transcription factors from Lambert et al and Human Protein Atlas annotation.
        - proteinatlas-TFs.tsv: all proteins annotated as transcription factors in the Human Protein Atlas
        - proteinatlas-cluster34.tsv: all proteins in cell-specific expression cluster 34 (same as MED12L)
        - cofactors_mapped.tsv: Uniprot mapping tool results for manually curated cofactor subunits
        - cofactors_mapped_combined.tsv: Uniprot mapping and paralog/cofactor complex annotation for each subunit. cofactors_mapped.tsv merged with cofactor_complexes>data>cofactor_complexes_manual.xlsx
        - tfsmapped_gencodes.csv: Uniprot mapping for TFs from TF-AD-RD>tfs_combined.csv
        - TFs.csv: list of EnsemblIDs for transcription factors, used in mapping

    - humanproteinatlas_mining.ipynb: 
        - API call to get scRNA-seq and Tau specificity scores from Human Protein Atlas
        - API call to get annotations from Human Protein Atlas
        - Application of API calls for cofactor subunits, ADs, RDs, transcription factors
        - Resulting .csv files are put in helperdata

    - humanproteinatlas_analysis.ipynb:
        - Looking at cell-specificity and co-expression of paralogous cofactor subunits
        - Analysis of transcription factor co-expression

    - mediator_exploration.ipynb
        - Cell-specificity of mediator subunits
        - Identifying co-expressed transcription factors
        - Annotations of co-expressed transcription factors
        - Identifying co-expressed AD/RDs
        - Exploring Cluster 34 (MED12L similar) proteins

    - BAFexploration.ipynb
        - Cell-specificity of BAF subunits

    - BAF_exploration2.ipynb
        - Cell-specificity of BAF subunits in more detail
        - Measuring co-expression between own-complex BAF paralogs vs all BAF pairs 

    - TFspecificity-old.ipynb
        - Cell-specificity and co-expression of full TF list, and AD/RDs
        - Sequence motifs, compositional bias, etc. (from Uniprot annotations) of co-expressed TFs 

mediator
    - data: sequences of Mediator subunits, Foldseek results, ALBATROSS results, metapredict results
    - phylo_raw: Phylo scores of MED12/12L by base pair from UCSC genome browser 
    - ..._exon_positions.csv: exon coordinates for MED12 and 12L genes
    - MED12_12L_structure.ipynb:
        - Map disordered regions of MED12 and MED12L
        - Sequence comparison and alignment of disordered regions
        - Sequence feature comparison of disordered regions
        - Manually-selected motif counting and comparison 
        - ALBATROSS comparison for tail IDR

finches
    - virtual environment to fun FINCHES

finches_IDRs
    - data: 
        - *.fasta: protein sequences for cofactors of interest
        - *_finches.csv: FINCHES results for different AD:x pairs
        - *_library.csv or *_AD.csv: AD sequences and activity information
        - mediator: csv files with MED12 and MED12L similar TF sequences
    -helperdata:
        - *_finches.csv: FINCHES results for different AD:x pairs. These were generated by finches.ipynb
        - *_ADs.csv: AD sequences
        - coAs.csv: manually selected list of 21 coactivator regions, loosely based on DelRosso 2024 paper
        - soto_cofactors_regions.csv: sequences of all Uniprot-annotated regions of proteins know to interact with Soto TFs (via BioGRID)
        - manual coAs: FINCHES results for AD:coAs, where the coAs are from coAs.csv
        - savioresults: FINCHES results for AD:coAs that were run using Savio, using the soto_cofactors_regions.csv cofactors domains
    - finches.ipynb: notebook for running FINCHES locally 
    - finches_sequencecleaning.ipynb: notebook used to get Soto AD, RDs and cofactor domains (soto_cofactors.csv, soto_cofactors_regions.csv)
    - AD_coA_analysis.ipynb: primary notebook for general analysis of FINCHES results for AD:coA pairs
        - Staller 2022 Predicted AD library with Soto Cofactor regions
            - AD hierarchical clustering and heatmap
            - FINCHES scores vs AD activity
            - coA hierarchical clustering and heatmap
            - AD:MED coA domains
        - Manual list of coAs:
            - AD:coA attractive vs repulsive epsilon scores with activity, for each coA 
            - Staller 2022 5 TF library AD attractive epsilon vs coAs 
            - FINCHES correlation between coAs
            - Testing ensembl regression models to predict activation from FINCHES scores
            - Epsilon score distribution across manual coAs for each library
            - Correlation between AD:coA FINCHES vs AD activity for each coA
            - Difference in AD:coA FINCHES for each AD vs AD activity
            - Exploration of only activating ADs
            - Plotting net charge of AD vs AD:coA FINCHES Epsilon for each coA
        - Running models:
            - ML regression models predicting AD activity from FINCHES binding profile
            - Clustering and PCA for each AD library, looking at activity differences across clusters
    - staller5lib_analysis.ipynb: in-depth look at FINCHES results for Staller 2022 5 Library TFs
        - AD:coA (coAs from soto_cofactors_regions.csv):
            - Random forest regression model to predict activity from AD binding profile
            - Looking at sequence properties for the most predictive coAs
            - Random forest regression model using attractive epsilon scores only
        - AD:AD, AD:coA combined:
            - Visualizing AD:AD self-interactions
            - Incorporation AD:AD scores into AD:coA-based activity prediction model (AD:coA - AD:AD)
        - AD:DBD FINCHES results, using synthetic DBD sequence from Staller 2022
    - mediatorcoexpression_ADcoA.ipynb: check whether co-expressed MED12/12L & TF ADs are more attractive
        - Prepare file for running FINCHES on MED12:ADs and MED12L:ADs
        - FINCHES score vs co-expression distance from MED12/L for IDRs and whole sequence
    - DelRosso: trying to recreate DelRosso 2024 AD:coA binding affinity results with FINCHES
        - delrosso_analysis.ipynb: 
            - Data cleaning
            - Compare FINCHES AD:coA profile vs DelRosso ddG 
            - Assess across each coA
            - Compare DelRosso binary binding results vs a FINCHES-based classifier, generate ROC and CM
            - Test classifier for different cutoff values