{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "import sys\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EnsemblId</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>TFAP2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000008196</td>\n",
       "      <td>TFAP2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000087510</td>\n",
       "      <td>TFAP2C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000008197</td>\n",
       "      <td>TFAP2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000116819</td>\n",
       "      <td>TFAP2E</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         EnsemblId  symbol\n",
       "1  ENSG00000137203  TFAP2A\n",
       "2  ENSG00000008196  TFAP2B\n",
       "3  ENSG00000087510  TFAP2C\n",
       "4  ENSG00000008197  TFAP2D\n",
       "5  ENSG00000116819  TFAP2E"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfs = pd.read_csv(\"original_data/lambert_supp_tables.csv\")\n",
    "tfs = tfs[tfs[\"Is TF?\"] == \"Yes\"][[\"EnsemblId\", \"symbol\"]]\n",
    "tfs.to_csv(\"lambert_tfs.csv\")\n",
    "tfs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>uniprotID</th>\n",
       "      <th>Matching Isoforms</th>\n",
       "      <th>Canonical Transcript ID</th>\n",
       "      <th>orig_uniprotID</th>\n",
       "      <th>Reference</th>\n",
       "      <th>TileType</th>\n",
       "      <th>ProteinRegionSeq</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABRAXAS1</td>\n",
       "      <td>121</td>\n",
       "      <td>200</td>\n",
       "      <td>Q6UWZ7</td>\n",
       "      <td>[['Q6UWZ7']]</td>\n",
       "      <td>ENST00000321945</td>\n",
       "      <td>Q6UWZ7</td>\n",
       "      <td>DelRosso et al.</td>\n",
       "      <td>CR</td>\n",
       "      <td>LQEHFSNQDLVFLLLTPSIITESCSTHRLEHSLYKPQKGLFHRVPL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AEBP1</td>\n",
       "      <td>1088</td>\n",
       "      <td>1158</td>\n",
       "      <td>Q8IUX7</td>\n",
       "      <td>[['Q8IUX7']]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q8IUX7</td>\n",
       "      <td>Staller Activity Data</td>\n",
       "      <td>TF</td>\n",
       "      <td>EVVTEFGTEVEPEFGTKVEPEFETQLEPEFETQLEPEFEEEEEEEK...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AHCTF1</td>\n",
       "      <td>1445</td>\n",
       "      <td>1698</td>\n",
       "      <td>Q8WYP5</td>\n",
       "      <td>[['Q8WYP5'], ['Q8WYP5']]</td>\n",
       "      <td>nan / nan</td>\n",
       "      <td>Q8WYP5 / Q8WYP5</td>\n",
       "      <td>PMID: 11952839, Soto / transcriptionalactivity...</td>\n",
       "      <td>TF</td>\n",
       "      <td>IRANDNKSMADVLGDGGNSSLTISEGPIVSERRLNQEVALNLKEDH...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AHR</td>\n",
       "      <td>118</td>\n",
       "      <td>126</td>\n",
       "      <td>P35869</td>\n",
       "      <td>[['P35869']]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P35869</td>\n",
       "      <td>transcriptionalactivity_regions.txt, GSL</td>\n",
       "      <td>TF</td>\n",
       "      <td>LLQALNGFV</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AHR</td>\n",
       "      <td>266</td>\n",
       "      <td>268</td>\n",
       "      <td>P35869</td>\n",
       "      <td>[['P35869']]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P35869</td>\n",
       "      <td>transcriptionalactivity_regions.txt, GSL</td>\n",
       "      <td>TF</td>\n",
       "      <td>FAI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene  Start   End uniprotID         Matching Isoforms  \\\n",
       "0  ABRAXAS1    121   200    Q6UWZ7              [['Q6UWZ7']]   \n",
       "1     AEBP1   1088  1158    Q8IUX7              [['Q8IUX7']]   \n",
       "2    AHCTF1   1445  1698    Q8WYP5  [['Q8WYP5'], ['Q8WYP5']]   \n",
       "3       AHR    118   126    P35869              [['P35869']]   \n",
       "4       AHR    266   268    P35869              [['P35869']]   \n",
       "\n",
       "  Canonical Transcript ID   orig_uniprotID  \\\n",
       "0         ENST00000321945           Q6UWZ7   \n",
       "1                     NaN           Q8IUX7   \n",
       "2               nan / nan  Q8WYP5 / Q8WYP5   \n",
       "3                     NaN           P35869   \n",
       "4                     NaN           P35869   \n",
       "\n",
       "                                           Reference TileType  \\\n",
       "0                                    DelRosso et al.       CR   \n",
       "1                              Staller Activity Data       TF   \n",
       "2  PMID: 11952839, Soto / transcriptionalactivity...       TF   \n",
       "3           transcriptionalactivity_regions.txt, GSL       TF   \n",
       "4           transcriptionalactivity_regions.txt, GSL       TF   \n",
       "\n",
       "                                    ProteinRegionSeq Notes  \n",
       "0  LQEHFSNQDLVFLLLTPSIITESCSTHRLEHSLYKPQKGLFHRVPL...   NaN  \n",
       "1  EVVTEFGTEVEPEFGTKVEPEFETQLEPEFETQLEPEFEEEEEEEK...   NaN  \n",
       "2  IRANDNKSMADVLGDGGNSSLTISEGPIVSERRLNQEVALNLKEDH...   NaN  \n",
       "3                                          LLQALNGFV   NaN  \n",
       "4                                                FAI   NaN  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ads = pd.read_csv(\"original_data/known_ADs_considering_isoforms_and_canonical.csv\")\n",
    "ads.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniprotId</th>\n",
       "      <th>DBDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A024R0Y4</td>\n",
       "      <td>[(11, 90)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A087X169</td>\n",
       "      <td>[(21, 140)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0A0MRP6</td>\n",
       "      <td>[(11, 140), (561, 650), (881, 960)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0B4J2G0</td>\n",
       "      <td>[(2, 81), (352, 431), (612, 731)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0C4DFL8</td>\n",
       "      <td>[(581, 700), (831, 910), (931, 1050)]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    UniprotId                                   DBDs\n",
       "0  A0A024R0Y4                             [(11, 90)]\n",
       "1  A0A087X169                            [(21, 140)]\n",
       "2  A0A0A0MRP6    [(11, 140), (561, 650), (881, 960)]\n",
       "3  A0A0B4J2G0      [(2, 81), (352, 431), (612, 731)]\n",
       "4  A0A0C4DFL8  [(581, 700), (831, 910), (931, 1050)]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rds = pd.read_csv(\"original_data/tycko_soto_delrosso_RD_coordinate_data.csv\")\n",
    "rds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "      <th>Tissue specificity</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Protein families</th>\n",
       "      <th>Compositional bias</th>\n",
       "      <th>Sequence similarities</th>\n",
       "      <th>Bgee</th>\n",
       "      <th>ExpressionAtlas</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>P05549</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>AP2A_HUMAN</td>\n",
       "      <td>Transcription factor AP-2-alpha (AP2-alpha) (A...</td>\n",
       "      <td>TFAP2A AP2TF TFAP2</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>437</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MOTIF 57..62; /note=\"PPxY motif\"</td>\n",
       "      <td>AP-2 family</td>\n",
       "      <td>COMPBIAS 23..48; /note=\"Polar residues\"; /evid...</td>\n",
       "      <td>SIMILARITY: Belongs to the AP-2 family. {ECO:0...</td>\n",
       "      <td>ENSG00000137203; Expressed in upper leg skin a...</td>\n",
       "      <td>P05549; baseline and differential.;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>A0A6E1XE14</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>A0A6E1XE14_HUMAN</td>\n",
       "      <td>Transcription factor AP-2-alpha (AP-2 transcri...</td>\n",
       "      <td>TFAP2A hCG_37170</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>439</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AP-2 family</td>\n",
       "      <td>COMPBIAS 25..50; /note=\"Polar residues\"; /evid...</td>\n",
       "      <td>SIMILARITY: Belongs to the AP-2 family. {ECO:0...</td>\n",
       "      <td>ENSG00000137203; Expressed in upper leg skin a...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>C1K3N0</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>C1K3N0_HUMAN</td>\n",
       "      <td>Transcription factor AP-2-alpha (AP-2 transcri...</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AP-2 family</td>\n",
       "      <td>COMPBIAS 23..48; /note=\"Polar residues\"; /evid...</td>\n",
       "      <td>SIMILARITY: Belongs to the AP-2 family. {ECO:0...</td>\n",
       "      <td>ENSG00000137203; Expressed in upper leg skin a...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>C9J6N8</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>C9J6N8_HUMAN</td>\n",
       "      <td>Transcription factor AP-2 alpha</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COMPBIAS 7..22; /note=\"Polar residues\"; /evide...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000137203; Expressed in upper leg skin a...</td>\n",
       "      <td>C9J6N8; baseline and differential.;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>C9JXZ2</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>C9JXZ2_HUMAN</td>\n",
       "      <td>Transcription factor AP-2 alpha</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>136</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AP-2 family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIMILARITY: Belongs to the AP-2 family. {ECO:0...</td>\n",
       "      <td>ENSG00000137203; Expressed in upper leg skin a...</td>\n",
       "      <td>C9JXZ2; baseline and differential.;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              From       Entry    Reviewed        Entry Name  \\\n",
       "0  ENSG00000137203      P05549    reviewed        AP2A_HUMAN   \n",
       "1  ENSG00000137203  A0A6E1XE14  unreviewed  A0A6E1XE14_HUMAN   \n",
       "2  ENSG00000137203      C1K3N0  unreviewed      C1K3N0_HUMAN   \n",
       "3  ENSG00000137203      C9J6N8  unreviewed      C9J6N8_HUMAN   \n",
       "4  ENSG00000137203      C9JXZ2  unreviewed      C9JXZ2_HUMAN   \n",
       "\n",
       "                                       Protein names          Gene Names  \\\n",
       "0  Transcription factor AP-2-alpha (AP2-alpha) (A...  TFAP2A AP2TF TFAP2   \n",
       "1  Transcription factor AP-2-alpha (AP-2 transcri...    TFAP2A hCG_37170   \n",
       "2  Transcription factor AP-2-alpha (AP-2 transcri...              TFAP2A   \n",
       "3                    Transcription factor AP-2 alpha              TFAP2A   \n",
       "4                    Transcription factor AP-2 alpha              TFAP2A   \n",
       "\n",
       "               Organism  Length Tissue specificity  \\\n",
       "0  Homo sapiens (Human)     437                NaN   \n",
       "1  Homo sapiens (Human)     439                NaN   \n",
       "2  Homo sapiens (Human)     328                NaN   \n",
       "3  Homo sapiens (Human)     173                NaN   \n",
       "4  Homo sapiens (Human)     136                NaN   \n",
       "\n",
       "                              Motif Protein families  \\\n",
       "0  MOTIF 57..62; /note=\"PPxY motif\"      AP-2 family   \n",
       "1                               NaN      AP-2 family   \n",
       "2                               NaN      AP-2 family   \n",
       "3                               NaN              NaN   \n",
       "4                               NaN      AP-2 family   \n",
       "\n",
       "                                  Compositional bias  \\\n",
       "0  COMPBIAS 23..48; /note=\"Polar residues\"; /evid...   \n",
       "1  COMPBIAS 25..50; /note=\"Polar residues\"; /evid...   \n",
       "2  COMPBIAS 23..48; /note=\"Polar residues\"; /evid...   \n",
       "3  COMPBIAS 7..22; /note=\"Polar residues\"; /evide...   \n",
       "4                                                NaN   \n",
       "\n",
       "                               Sequence similarities  \\\n",
       "0  SIMILARITY: Belongs to the AP-2 family. {ECO:0...   \n",
       "1  SIMILARITY: Belongs to the AP-2 family. {ECO:0...   \n",
       "2  SIMILARITY: Belongs to the AP-2 family. {ECO:0...   \n",
       "3                                                NaN   \n",
       "4  SIMILARITY: Belongs to the AP-2 family. {ECO:0...   \n",
       "\n",
       "                                                Bgee  \\\n",
       "0  ENSG00000137203; Expressed in upper leg skin a...   \n",
       "1  ENSG00000137203; Expressed in upper leg skin a...   \n",
       "2  ENSG00000137203; Expressed in upper leg skin a...   \n",
       "3  ENSG00000137203; Expressed in upper leg skin a...   \n",
       "4  ENSG00000137203; Expressed in upper leg skin a...   \n",
       "\n",
       "                       ExpressionAtlas  \n",
       "0  P05549; baseline and differential.;  \n",
       "1                                  NaN  \n",
       "2                                  NaN  \n",
       "3  C9J6N8; baseline and differential.;  \n",
       "4  C9JXZ2; baseline and differential.;  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfsmapped = pd.read_csv(\"maps/ID Mapping Dec 10 2024.tsv\", delimiter=\"\\t\")\n",
    "tfsmapped.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "      <th>Tissue specificity</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Protein families</th>\n",
       "      <th>Compositional bias</th>\n",
       "      <th>Sequence similarities</th>\n",
       "      <th>Bgee</th>\n",
       "      <th>ExpressionAtlas</th>\n",
       "      <th>num ADs</th>\n",
       "      <th>num RDs</th>\n",
       "      <th>classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>P05549</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>AP2A_HUMAN</td>\n",
       "      <td>Transcription factor AP-2-alpha (AP2-alpha) (A...</td>\n",
       "      <td>TFAP2A AP2TF TFAP2</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>437</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MOTIF 57..62; /note=\"PPxY motif\"</td>\n",
       "      <td>AP-2 family</td>\n",
       "      <td>COMPBIAS 23..48; /note=\"Polar residues\"; /evid...</td>\n",
       "      <td>SIMILARITY: Belongs to the AP-2 family. {ECO:0...</td>\n",
       "      <td>ENSG00000137203; Expressed in upper leg skin a...</td>\n",
       "      <td>P05549; baseline and differential.;</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>AD only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>A0A6E1XE14</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>A0A6E1XE14_HUMAN</td>\n",
       "      <td>Transcription factor AP-2-alpha (AP-2 transcri...</td>\n",
       "      <td>TFAP2A hCG_37170</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>439</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AP-2 family</td>\n",
       "      <td>COMPBIAS 25..50; /note=\"Polar residues\"; /evid...</td>\n",
       "      <td>SIMILARITY: Belongs to the AP-2 family. {ECO:0...</td>\n",
       "      <td>ENSG00000137203; Expressed in upper leg skin a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>C1K3N0</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>C1K3N0_HUMAN</td>\n",
       "      <td>Transcription factor AP-2-alpha (AP-2 transcri...</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AP-2 family</td>\n",
       "      <td>COMPBIAS 23..48; /note=\"Polar residues\"; /evid...</td>\n",
       "      <td>SIMILARITY: Belongs to the AP-2 family. {ECO:0...</td>\n",
       "      <td>ENSG00000137203; Expressed in upper leg skin a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>C9J6N8</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>C9J6N8_HUMAN</td>\n",
       "      <td>Transcription factor AP-2 alpha</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COMPBIAS 7..22; /note=\"Polar residues\"; /evide...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000137203; Expressed in upper leg skin a...</td>\n",
       "      <td>C9J6N8; baseline and differential.;</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>C9JXZ2</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>C9JXZ2_HUMAN</td>\n",
       "      <td>Transcription factor AP-2 alpha</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>136</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AP-2 family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIMILARITY: Belongs to the AP-2 family. {ECO:0...</td>\n",
       "      <td>ENSG00000137203; Expressed in upper leg skin a...</td>\n",
       "      <td>C9JXZ2; baseline and differential.;</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              From       Entry    Reviewed        Entry Name  \\\n",
       "0  ENSG00000137203      P05549    reviewed        AP2A_HUMAN   \n",
       "1  ENSG00000137203  A0A6E1XE14  unreviewed  A0A6E1XE14_HUMAN   \n",
       "2  ENSG00000137203      C1K3N0  unreviewed      C1K3N0_HUMAN   \n",
       "3  ENSG00000137203      C9J6N8  unreviewed      C9J6N8_HUMAN   \n",
       "4  ENSG00000137203      C9JXZ2  unreviewed      C9JXZ2_HUMAN   \n",
       "\n",
       "                                       Protein names          Gene Names  \\\n",
       "0  Transcription factor AP-2-alpha (AP2-alpha) (A...  TFAP2A AP2TF TFAP2   \n",
       "1  Transcription factor AP-2-alpha (AP-2 transcri...    TFAP2A hCG_37170   \n",
       "2  Transcription factor AP-2-alpha (AP-2 transcri...              TFAP2A   \n",
       "3                    Transcription factor AP-2 alpha              TFAP2A   \n",
       "4                    Transcription factor AP-2 alpha              TFAP2A   \n",
       "\n",
       "               Organism  Length Tissue specificity  \\\n",
       "0  Homo sapiens (Human)     437                NaN   \n",
       "1  Homo sapiens (Human)     439                NaN   \n",
       "2  Homo sapiens (Human)     328                NaN   \n",
       "3  Homo sapiens (Human)     173                NaN   \n",
       "4  Homo sapiens (Human)     136                NaN   \n",
       "\n",
       "                              Motif Protein families  \\\n",
       "0  MOTIF 57..62; /note=\"PPxY motif\"      AP-2 family   \n",
       "1                               NaN      AP-2 family   \n",
       "2                               NaN      AP-2 family   \n",
       "3                               NaN              NaN   \n",
       "4                               NaN      AP-2 family   \n",
       "\n",
       "                                  Compositional bias  \\\n",
       "0  COMPBIAS 23..48; /note=\"Polar residues\"; /evid...   \n",
       "1  COMPBIAS 25..50; /note=\"Polar residues\"; /evid...   \n",
       "2  COMPBIAS 23..48; /note=\"Polar residues\"; /evid...   \n",
       "3  COMPBIAS 7..22; /note=\"Polar residues\"; /evide...   \n",
       "4                                                NaN   \n",
       "\n",
       "                               Sequence similarities  \\\n",
       "0  SIMILARITY: Belongs to the AP-2 family. {ECO:0...   \n",
       "1  SIMILARITY: Belongs to the AP-2 family. {ECO:0...   \n",
       "2  SIMILARITY: Belongs to the AP-2 family. {ECO:0...   \n",
       "3                                                NaN   \n",
       "4  SIMILARITY: Belongs to the AP-2 family. {ECO:0...   \n",
       "\n",
       "                                                Bgee  \\\n",
       "0  ENSG00000137203; Expressed in upper leg skin a...   \n",
       "1  ENSG00000137203; Expressed in upper leg skin a...   \n",
       "2  ENSG00000137203; Expressed in upper leg skin a...   \n",
       "3  ENSG00000137203; Expressed in upper leg skin a...   \n",
       "4  ENSG00000137203; Expressed in upper leg skin a...   \n",
       "\n",
       "                       ExpressionAtlas  num ADs  num RDs classification  \n",
       "0  P05549; baseline and differential.;        1        0        AD only  \n",
       "1                                  NaN        0        0           none  \n",
       "2                                  NaN        0        0           none  \n",
       "3  C9J6N8; baseline and differential.;        0        0           none  \n",
       "4  C9JXZ2; baseline and differential.;        0        0           none  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def ad_rd_annotate(uniprot):\n",
    "    numads = len(ads[ads[\"uniprotID\"] == uniprot])\n",
    "    numrds = len(rds[rds[\"UniprotId\"] == uniprot])\n",
    "    if numads + numrds == 0:\n",
    "        classification = \"none\"\n",
    "    elif (numads > 0) and (numrds > 0):\n",
    "        classification = \"both\"\n",
    "    elif numads > 0:\n",
    "        classification = \"AD only\"\n",
    "    elif numrds > 0:\n",
    "        classification = \"RD only\"\n",
    "    return numads, numrds, classification\n",
    "\n",
    "numadslist = []\n",
    "numrdslist = []\n",
    "classificationlist = []\n",
    "for uniprot in tfsmapped[\"Entry\"]:\n",
    "    numads, numrds, classification = ad_rd_annotate(uniprot)\n",
    "    numadslist.append(numads)\n",
    "    numrdslist.append(numrds)\n",
    "    classificationlist.append(classification)\n",
    "\n",
    "tfsmapped[\"num ADs\"] = numadslist\n",
    "tfsmapped[\"num RDs\"] = numrdslist\n",
    "tfsmapped[\"classification\"] = classificationlist\n",
    "\n",
    "tfsmapped.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "classification\n",
       "none       5010\n",
       "RD only     701\n",
       "both        369\n",
       "AD only      90\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfsmapped[\"classification\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%skip\n",
    "def gtex_getgenecode(ensemblid):\n",
    "    server = \"https://gtexportal.org/api/v2/reference/gene?\"\n",
    "    ext = \"geneId=\" + ensemblid\n",
    "    r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        sys.exit()\n",
    "    \n",
    "    decoded = r.json()\n",
    "\n",
    "    return decoded[\"data\"][0][\"gencodeId\"]\n",
    "\n",
    "tfsmapped[\"gencodeId\"] = tfsmapped[\"From\"].apply(gtex_getgenecode)\n",
    "tfsmapped.to_csv(\"tfsmapped-gencodes.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000267281.2error\n",
      "ENSG00000267179.1error\n",
      "ENSG00000264668.1error\n",
      "ENSG00000100146.16error\n",
      "ENSG00000064489.22error\n"
     ]
    }
   ],
   "source": [
    "%%skip\n",
    "def gtex_fullexpression(ensemblid):\n",
    "    server = \"https://gtexportal.org/api/v2/expression/geneExpression?\"\n",
    "    ext = \"gencodeId=\" + ensemblid\n",
    "    r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        sys.exit()\n",
    "    \n",
    "    decoded = r.json()\n",
    "\n",
    "    return pd.DataFrame(decoded[\"data\"])\n",
    "\n",
    "\n",
    "def gtex_fullexpression_clean(ensemblid, save_csv, dirstring):\n",
    "    \"\"\"\n",
    "    Ensemblid: \n",
    "    Save_csv (boolean): \n",
    "    dirstring (string): directory to save the csv into\n",
    "    value (string): 1D df of mean, median, std, count from the expression results to return\n",
    "    \"\"\"\n",
    "    fulldf = gtex_fullexpression(ensemblid)\n",
    "\n",
    "    newdf = fulldf[[\"tissueSiteDetailId\"]].set_index(\"tissueSiteDetailId\")\n",
    "    newdf[\"mean\"] = fulldf[\"data\"].apply(np.mean).to_list()\n",
    "    newdf[\"median\"] = fulldf[\"data\"].apply(np.mean).to_list()\n",
    "    newdf[\"std\"] = fulldf[\"data\"].apply(np.std).to_list()\n",
    "    newdf[\"count\"] = fulldf[\"data\"].apply(len).to_list()\n",
    "    \n",
    "    def select_df(label):\n",
    "        return newdf[[label]].rename({label: ensemblid}, axis=1)\n",
    "\n",
    "    if save_csv:\n",
    "        newdf.to_csv(dirstring + ensemblid + \"_expression.csv\")\n",
    "\n",
    "    return select_df(\"mean\"), select_df(\"median\"), select_df(\"std\"), select_df(\"count\")\n",
    "\n",
    "\n",
    "################## Run on TFs\n",
    "\n",
    "means, medians, stds, counts = pd.DataFrame({}), pd.DataFrame({}), pd.DataFrame({}), pd.DataFrame({})\n",
    "\n",
    "for ensemblId in tfsmapped[\"gencodeId\"].unique():\n",
    "    try:\n",
    "        mean_i, median_i, std_i, count_i = gtex_fullexpression_clean(ensemblId, False, \"TF_expression/\")\n",
    "        means = means.join(mean_i, how = \"outer\")\n",
    "        medians = medians.join(median_i, how = \"outer\")\n",
    "        stds = stds.join(std_i, how = \"outer\")\n",
    "        counts = counts.join(count_i, how = \"outer\")\n",
    "    except:\n",
    "        print(ensemblId + \"error\")\n",
    "\n",
    "means.to_csv(\"ensemblId_mean.csv\")\n",
    "medians.to_csv(\"ensemblId_median.csv\")\n",
    "stds.to_csv(\"ensemblId_stds.csv\")\n",
    "counts.to_csv(\"ensemblId_counts.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
